runtime/cudaq/platform/default/rest/helpers/anyon/AnyonServerHelper.cpp

```diff
@@ -0,0 +1,464 @@
+    apiKey = response_json["id_token"].get<std::string>();
+    refreshKey = response_json["refresh_token"].get<std::string>();
+    std::ofstream out(credentialsPath);
+    out << "key:" << apiKey << '\n';
+    out << "refresh:" << refreshKey << '\n';
+    out << "time:" << now.time_since_epoch().count() << '\n';
+    timeStr = std::to_string(now.time_since_epoch().count());
+  }
+  // If the time string is empty, let's add it
+  if (timeStr.empty()) {
+    timeStr = std::to_string(now.time_since_epoch().count());
+    std::ofstream out(credentialsPath);
+    out << "key:" << apiKey << '\n';
+    out << "refresh:" << refreshKey << '\n';
+    out << "time:" << timeStr << '\n';
+  }
+}
+
+void findApiKeyInFile(std::string &apiKey, const std::string &path,
+                      std::string &refreshKey, std::string &timeStr,
+                      std::string &credentials) {
+  std::ifstream stream(path);
+  std::string contents((std::istreambuf_iterator<char>(stream)),
+                       std::istreambuf_iterator<char>());
+
+  std::vector<std::string> lines;
+  lines = cudaq::split(contents, '\n');
+  nlohmann::json jsoncreds;
+  for (const std::string &l : lines) {
+    std::vector<std::string> keyAndValue = cudaq::split(l, ':');
+    // if (keyAndValue.size() != 2)
+    //   throw std::runtime_error("Ill-formed configuration file (" + path +
+    //                            "). Key-value pairs must be in `<key> : "
+    //                            "<value>` format. (One per line)");
+    cudaq::trim(keyAndValue[0]);
+    cudaq::trim(keyAndValue[1]);
+    if (keyAndValue[0] == "key")
+      apiKey = keyAndValue[1];
+    else if (keyAndValue[0] == "refresh")
+      refreshKey = keyAndValue[1];
+    else if (keyAndValue[0] == "time")
+      timeStr = keyAndValue[1];
+    else if (keyAndValue[0] ==
+             "credentials") { // If the config file doesn't contain key and
+                              // refresh token, we will add the username
+                              // password to apikey for BasicHttpAuthentication
+                              // and generation of tokens
+      std::string linecontent =
+          keyAndValue[1] + ":" + keyAndValue[2] + ":" + keyAndValue[3];
+      // printf("The credentials read from the .config file is: %s",
+      // linecontent.c_str());
+      jsoncreds = json::parse(linecontent);
+      std::string delim(":");
+      std::string username = jsoncreds.at("username");
+      std::string passwd = jsoncreds.at("password");
+      std::string authInfo = username + delim + passwd;
+      // authInfo = base64::to_base64(authInfo);
+      authInfo = llvm::encodeBase64(authInfo);
+      credentials = "Basic " + authInfo;
+    } else
+      throw std::runtime_error(
+          "Unknown key in configuration file: " + keyAndValue[0] + ".");
+  }
+  // if (apiKey.empty())
+  //   throw std::runtime_error("Empty API key in configuration file (" + path +
+  //                            ").");
```

Comment:
Is there any particular reason this is commented out? If not, could you please uncomment it?

Reply:
the api key can be empty as long as user provides "credentials: {username: <username>, password:<password>}" or "refresh_key: <refresh_key>". The lines below this segment capture this exception. The Helper will automatically work with remote REST server to get authenticated through the username/password or existing refresh_key. If none of these exist, an exception will be thrown. We assumed that API_key is routinely discarded/updated while refresh_key is a little more persistent and username/password is "permanent". API_key is used for constant communication and authentication with remote REST server. A valid refresh_key is used to request new api_key when api_key is absent or expired. Username/Password is used to get both in the beginning and should be avoided as much as possible.

Reply:
just implemented the change